diff --git a/src/com/google/javascript/jscomp/GlobalNamespace.java b/src/com/google/javascript/jscomp/GlobalNamespace.java
index 639318df..0e9c4d33 100644
--- a/src/com/google/javascript/jscomp/GlobalNamespace.java
+++ b/src/com/google/javascript/jscomp/GlobalNamespace.java
@@ -363,6 +363,7 @@ class GlobalNamespace
                 isSet = true;
                 type = Name.Type.FUNCTION;
                 break;
+              case Token.CATCH:
               case Token.INC:
               case Token.DEC:
                 isSet = true;
diff --git a/test/com/google/javascript/jscomp/CheckGlobalNamesTest.java b/test/com/google/javascript/jscomp/CheckGlobalNamesTest.java
index 1050872c..46152b9d 100644
--- a/test/com/google/javascript/jscomp/CheckGlobalNamesTest.java
+++ b/test/com/google/javascript/jscomp/CheckGlobalNamesTest.java
@@ -336,4 +336,13 @@ public class CheckGlobalNamesTest extends CompilerTestCase {
         "SubFoo.superClazz();",
          UNDEFINED_NAME_WARNING);
   }
+
+  public void testGlobalCatch() throws Exception {
+    testSame(
+        "try {" +
+        "  throw Error();" +
+        "} catch (e) {" +
+        "  console.log(e.name)" +
+        "}");
+  }
 }
diff --git a/test/com/google/javascript/jscomp/CollapsePropertiesTest.java b/test/com/google/javascript/jscomp/CollapsePropertiesTest.java
index 51020876..4a812cea 100644
--- a/test/com/google/javascript/jscomp/CollapsePropertiesTest.java
+++ b/test/com/google/javascript/jscomp/CollapsePropertiesTest.java
@@ -1689,4 +1689,13 @@ public class CollapsePropertiesTest extends CompilerTestCase {
         "  log(namespace$Param$optParam);\n" +
         "}");
   }
+
+  public void testGlobalCatch() throws Exception {
+    testSame(
+        "try {" +
+        "  throw Error();" +
+        "} catch (e) {" +
+        "  console.log(e.name)" +
+        "}");
+  }
 }
