diff --git a/src/com/google/javascript/jscomp/CodeGenerator.java b/src/com/google/javascript/jscomp/CodeGenerator.java
index 3b2a9d42..8d268b6b 100644
--- a/src/com/google/javascript/jscomp/CodeGenerator.java
+++ b/src/com/google/javascript/jscomp/CodeGenerator.java
@@ -282,11 +282,12 @@ class CodeGenerator {
       case Token.HOOK: {
         Preconditions.checkState(childCount == 3);
         int p = NodeUtil.precedence(type);
+        Context rhsContext = getContextForNoInOperator(context);
         addExpr(first, p + 1, context);
         cc.addOp("?", true);
-        addExpr(first.getNext(), 1, Context.OTHER);
+        addExpr(first.getNext(), 1, rhsContext);
         cc.addOp(":", true);
-        addExpr(last, 1, Context.OTHER);
+        addExpr(last, 1, rhsContext);
         break;
       }
 
diff --git a/test/com/google/javascript/jscomp/CodePrinterTest.java b/test/com/google/javascript/jscomp/CodePrinterTest.java
index de53fc76..a369e4c6 100644
--- a/test/com/google/javascript/jscomp/CodePrinterTest.java
+++ b/test/com/google/javascript/jscomp/CodePrinterTest.java
@@ -466,6 +466,9 @@ public class CodePrinterTest extends TestCase {
     // Test we correctly handle an in operator inside a comma.
     assertPrintSame("for(x,(y in z);;)foo()");
     assertPrintSame("for(var x,w=(y in z);;)foo()");
+
+    // And in operator inside a hook.
+    assertPrintSame("for(a=c?0:(0 in d);;)foo()");
   }
 
   public void testLiteralProperty() {
