diff --git a/src/com/google/javascript/jscomp/InlineVariables.java b/src/com/google/javascript/jscomp/InlineVariables.java
index 319b61c3..7d99778c 100644
--- a/src/com/google/javascript/jscomp/InlineVariables.java
+++ b/src/com/google/javascript/jscomp/InlineVariables.java
@@ -196,6 +196,8 @@ class InlineVariables implements CompilerPass {
     private void doInlinesForScope(NodeTraversal t,
         Map<Var, ReferenceCollection> referenceMap) {
 
+      boolean maybeModifiedArguments =
+          maybeEscapedOrModifiedArguments(t.getScope(), referenceMap);
       for (Iterator<Var> it = t.getScope().getVars(); it.hasNext();) {
         Var v = it.next();
 
@@ -217,13 +219,44 @@ class InlineVariables implements CompilerPass {
           // inlining heuristics. See InlineConstantsTest.
           continue;
         } else {
-          inlineNonConstants(v, referenceInfo);
+          inlineNonConstants(v, referenceInfo, maybeModifiedArguments);
         }
       }
     }
 
+    private boolean maybeEscapedOrModifiedArguments(
+        Scope scope, Map<Var, ReferenceCollection> referenceMap) {
+      if (scope.isLocal()) {
+        Var arguments = scope.getArgumentsVar();
+        ReferenceCollection refs = referenceMap.get(arguments);
+        if (refs != null && !refs.references.isEmpty()) {
+          for (Reference ref : refs.references) {
+            Node refNode = ref.getNameNode();
+            Node refParent = ref.getParent();
+            // Any reference that is not a read of the arguments property
+            // consider a escape of the arguments object.
+            if (!(NodeUtil.isGet(refParent)
+                && refNode == ref.getParent().getFirstChild()
+                && !isLValue(refParent))) {
+              return true;
+            }
+          }
+        }
+      }
+      return false;
+    }
+
+    private boolean isLValue(Node n) {
+      Node parent = n.getParent();
+      return (parent.getType() == Token.INC
+          || parent.getType() == Token.DEC
+          || (NodeUtil.isAssignmentOp(parent)
+          && parent.getFirstChild() == n));
+    }
+
     private void inlineNonConstants(
-        Var v, ReferenceCollection referenceInfo) {
+        Var v, ReferenceCollection referenceInfo,
+        boolean maybeModifiedArguments) {
       int refCount = referenceInfo.references.size();
       Reference declaration = referenceInfo.references.get(0);
       Reference init = referenceInfo.getInitializingReference();
@@ -268,7 +301,8 @@ class InlineVariables implements CompilerPass {
       // inline an alias of it. (If the variable was inlined, then the
       // reference data is out of sync. We're better off just waiting for
       // the next pass.)
-      if (!staleVars.contains(v) && referenceInfo.isWellDefined() &&
+      if (!maybeModifiedArguments &&
+          !staleVars.contains(v) && referenceInfo.isWellDefined() &&
           referenceInfo.isAssignedOnceInLifetime()) {
         List<Reference> refs = referenceInfo.references;
         for (int i = 1 /* start from a read */; i < refs.size(); i++) {
diff --git a/src/com/google/javascript/jscomp/ReferenceCollectingCallback.java b/src/com/google/javascript/jscomp/ReferenceCollectingCallback.java
index e3edbb4d..4c00d60e 100644
--- a/src/com/google/javascript/jscomp/ReferenceCollectingCallback.java
+++ b/src/com/google/javascript/jscomp/ReferenceCollectingCallback.java
@@ -120,7 +120,12 @@ class ReferenceCollectingCallback implements ScopedCallback, CompilerPass {
    */
   public void visit(NodeTraversal t, Node n, Node parent) {
     if (n.getType() == Token.NAME) {
-      Var v = t.getScope().getVar(n.getString());
+      Var v;
+      if (n.getString().equals("arguments")) {
+        v = t.getScope().getArgumentsVar();
+      } else {
+        v = t.getScope().getVar(n.getString());
+      }
       if (v != null && varFilter.apply(v)) {
         addReference(t, v,
             new Reference(n, parent, t, blockStack.peek()));
diff --git a/src/com/google/javascript/jscomp/Scope.java b/src/com/google/javascript/jscomp/Scope.java
index f90abb0d..19ac9d2b 100644
--- a/src/com/google/javascript/jscomp/Scope.java
+++ b/src/com/google/javascript/jscomp/Scope.java
@@ -60,6 +60,8 @@ public class Scope implements StaticScope<JSType> {
   /** Whether this is a bottom scope for the purposes of type inference. */
   private final boolean isBottom;
 
+  private Var arguments;
+
   /** Stores info about a variable */
   public static class Var implements StaticSlot<JSType> {
     /** name */
@@ -174,7 +176,7 @@ public class Scope implements StaticScope<JSType> {
      * based on the value reported by {@code NodeUtil}.
      */
     public boolean isConst() {
-      return NodeUtil.isConstantName(nameNode);
+      return nameNode != null && NodeUtil.isConstantName(nameNode);
     }
 
     /**
@@ -281,6 +283,41 @@ public class Scope implements StaticScope<JSType> {
     }
   }
 
+  /**
+   * A special subclass of Var used to distinguish "arguments" in the current
+   * scope.
+   */
+  // TODO(johnlenz): Include this the list of Vars for the scope.
+  public static class Arguments extends Var {
+    Arguments(Scope scope) {
+      super(
+        false, // no inferred
+        "arguments", // always arguments
+        null,  // no declaration node
+        // TODO(johnlenz): provide the type of "Arguments".
+        null,  // no type info
+        scope,
+        -1,    // no variable index
+        null,  // input,
+        false, // not a define
+        null   // no jsdoc
+        );
+    }
+
+    @Override public boolean equals(Object other) {
+      if (!(other instanceof Arguments)) {
+        return false;
+      }
+
+      Arguments otherVar = (Arguments) other;
+      return otherVar.scope.getRootNode() == scope.getRootNode();
+    }
+
+    @Override public int hashCode() {
+      return System.identityHashCode(this);
+    }
+  }
+
   /**
    * Creates a Scope given the parent Scope and the root node of the scope.
    * @param parent  The parent Scope. Cannot be null.
@@ -442,6 +479,16 @@ public class Scope implements StaticScope<JSType> {
     }
   }
 
+  /**
+   * Get a unique VAR object to represents "arguments" within this scope
+   */
+  public Var getArgumentsVar() {
+    if (arguments == null) {
+      arguments = new Arguments(this);
+    }
+    return arguments;
+  }
+
   /**
    * Returns true if a variable is declared.
    */
diff --git a/test/com/google/javascript/jscomp/InlineVariablesTest.java b/test/com/google/javascript/jscomp/InlineVariablesTest.java
index 02e2bab9..e1796c3e 100644
--- a/test/com/google/javascript/jscomp/InlineVariablesTest.java
+++ b/test/com/google/javascript/jscomp/InlineVariablesTest.java
@@ -910,4 +910,146 @@ public class InlineVariablesTest extends CompilerTestCase {
   public void testInlineNamedFunction() {
     test("function f() {} f();", "(function f(){})()");
   }
+
+  public void testIssue378ModifiedArguments1() {
+    testSame(
+        "function g(callback) {\n" +
+        "  var f = callback;\n" +
+        "  arguments[0] = this;\n" +
+        "  f.apply(this, arguments);\n" +
+        "}");
+  }
+
+  public void testIssue378ModifiedArguments2() {
+    testSame(
+        "function g(callback) {\n" +
+        "  /** @const */\n" +
+        "  var f = callback;\n" +
+        "  arguments[0] = this;\n" +
+        "  f.apply(this, arguments);\n" +
+        "}");
+  }
+
+  public void testIssue378EscapedArguments1() {
+    testSame(
+        "function g(callback) {\n" +
+        "  var f = callback;\n" +
+        "  h(arguments,this);\n" +
+        "  f.apply(this, arguments);\n" +
+        "}\n" +
+        "function h(a,b) {\n" +
+        "  a[0] = b;" +
+        "}");
+  }
+
+  public void testIssue378EscapedArguments2() {
+    testSame(
+        "function g(callback) {\n" +
+        "  /** @const */\n" +
+        "  var f = callback;\n" +
+        "  h(arguments,this);\n" +
+        "  f.apply(this);\n" +
+        "}\n" +
+        "function h(a,b) {\n" +
+        "  a[0] = b;" +
+        "}");
+  }
+
+  public void testIssue378EscapedArguments3() {
+    test(
+        "function g(callback) {\n" +
+        "  var f = callback;\n" +
+        "  f.apply(this, arguments);\n" +
+        "}\n",
+        "function g(callback) {\n" +
+        "  callback.apply(this, arguments);\n" +
+        "}\n");
+  }
+
+  public void testIssue378EscapedArguments4() {
+    testSame(
+        "function g(callback) {\n" +
+        "  var f = callback;\n" +
+        "  h(arguments[0],this);\n" +
+        "  f.apply(this, arguments);\n" +
+        "}\n" +
+        "function h(a,b) {\n" +
+        "  a[0] = b;" +
+        "}");
+  }
+
+  public void testIssue378ArgumentsRead1() {
+    test(
+        "function g(callback) {\n" +
+        "  var f = callback;\n" +
+        "  var g = arguments[0];\n" +
+        "  f.apply(this, arguments);\n" +
+        "}",
+        "function g(callback) {\n" +
+        "  var g = arguments[0];\n" +
+        "  callback.apply(this, arguments);\n" +
+        "}");
+  }
+
+  public void testIssue378ArgumentsRead2() {
+    test(
+        "function g(callback) {\n" +
+        "  var f = callback;\n" +
+        "  h(arguments[0],this);\n" +
+        "  f.apply(this, arguments[0]);\n" +
+        "}\n" +
+        "function h(a,b) {\n" +
+        "  a[0] = b;" +
+        "}",
+        "function g(callback) {\n" +
+        "  h(arguments[0],this);\n" +
+        "  callback.apply(this, arguments[0]);\n" +
+        "}\n" +
+        "function h(a,b) {\n" +
+        "  a[0] = b;" +
+        "}");
+  }
+
+  public void testArgumentsModifiedInOuterFunction() {
+    test(
+      "function g(callback) {\n" +
+      "  var f = callback;\n" +
+      "  arguments[0] = this;\n" +
+      "  f.apply(this, arguments);\n" +
+      "  function inner(callback) {" +
+      "    var x = callback;\n" +
+      "    x.apply(this);\n" +
+      "  }" +
+      "}",
+      "function g(callback) {\n" +
+      "  var f = callback;\n" +
+      "  arguments[0] = this;\n" +
+      "  f.apply(this, arguments);\n" +
+      "  function inner(callback) {" +
+      "    callback.apply(this);\n" +
+      "  }" +
+      "}");
+  }
+
+  public void testArgumentsModifiedInInnerFunction() {
+    test(
+      "function g(callback) {\n" +
+      "  var f = callback;\n" +
+      "  f.apply(this, arguments);\n" +
+      "  function inner(callback) {" +
+      "    var x = callback;\n" +
+      "    arguments[0] = this;\n" +
+      "    x.apply(this);\n" +
+      "  }" +
+      "}",
+      "function g(callback) {\n" +
+      "  callback.apply(this, arguments);\n" +
+      "  function inner(callback) {" +
+      "    var x = callback;\n" +
+      "    arguments[0] = this;\n" +
+      "    x.apply(this);\n" +
+      "  }" +
+      "}");
+  }
+
 }
