diff --git a/build.xml b/build.xml
index adec5c33..7a31609c 100644
--- a/build.xml
+++ b/build.xml
@@ -163,7 +163,7 @@
         <pathelement location="${build.dir}/test" />
       </classpath>
       <formatter type="plain" usefile="false" />
-      <batchtest>
+      <batchtest haltonfailure="true">
         <fileset dir="${build.dir}/test">
           <include name="**/*Test.class" />
         </fileset>
diff --git a/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java b/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java
index c929a363..1d738b52 100644
--- a/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java
+++ b/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java
@@ -786,7 +786,7 @@ class ProcessClosurePrimitives extends AbstractPostOrderCallback
         } else {
           // In this case, the name was implicitly provided by two independent
           // modules. We need to move this code up to a common module.
-          int indexOfDot = namespace.indexOf('.');
+          int indexOfDot = namespace.lastIndexOf('.');
           if (indexOfDot == -1) {
             // Any old place is fine.
             compiler.getNodeForCodeInsertion(minimumModule)
diff --git a/test/com/google/javascript/jscomp/ProcessClosurePrimitivesTest.java b/test/com/google/javascript/jscomp/ProcessClosurePrimitivesTest.java
index 5be0a1f0..f392a6cc 100644
--- a/test/com/google/javascript/jscomp/ProcessClosurePrimitivesTest.java
+++ b/test/com/google/javascript/jscomp/ProcessClosurePrimitivesTest.java
@@ -771,6 +771,21 @@ public class ProcessClosurePrimitivesTest extends CompilerTestCase {
         });
   }
 
+  public void testProvideInIndependentModules4() {
+    // Regression test for bug 261:
+    // http://code.google.com/p/closure-compiler/issues/detail?id=261
+    test(
+        createModuleStar(
+            "goog.provide('apps');",
+            "goog.provide('apps.foo.bar.B');",
+            "goog.provide('apps.foo.bar.C');"),
+        new String[] {
+            "var apps = {};apps.foo = {};apps.foo.bar = {}",
+            "apps.foo.bar.B = {};",
+            "apps.foo.bar.C = {};",
+        });
+  }
+
   public void testRequireOfBaseGoog() {
     test("goog.require('goog');",
          "", MISSING_PROVIDE_ERROR);
