diff --git a/lib/google_common_deploy.jar b/lib/google_common_deploy.jar
index 25740509..d1029961 100755
Binary files a/lib/google_common_deploy.jar and b/lib/google_common_deploy.jar differ
diff --git a/src/com/google/javascript/jscomp/CodingConvention.java b/src/com/google/javascript/jscomp/CodingConvention.java
index 95403499..4416fc4c 100644
--- a/src/com/google/javascript/jscomp/CodingConvention.java
+++ b/src/com/google/javascript/jscomp/CodingConvention.java
@@ -28,7 +28,7 @@ import java.util.Map;
  * Compiler for a specific team/company.
  *
  * // TODO(bolinfest): Tighten up this interface -- it is far too big.
- *
+ * 
 *
 *
  */
@@ -241,16 +241,12 @@ public interface CodingConvention {
 
   static class SubclassRelationship {
     final SubclassType type;
-    final Node subclassNode;
-    final Node superclassNode;
     final String subclassName;
     final String superclassName;
 
     SubclassRelationship(SubclassType type,
         Node subclassNode, Node superclassNode) {
       this.type = type;
-      this.subclassNode = subclassNode;
-      this.superclassNode = superclassNode;
       this.subclassName = subclassNode.getQualifiedName();
       this.superclassName = superclassNode.getQualifiedName();
     }
diff --git a/src/com/google/javascript/jscomp/CommandLineRunner.java b/src/com/google/javascript/jscomp/CommandLineRunner.java
index bdf55921..c18639ac 100644
--- a/src/com/google/javascript/jscomp/CommandLineRunner.java
+++ b/src/com/google/javascript/jscomp/CommandLineRunner.java
@@ -430,10 +430,8 @@ public class CommandLineRunner extends
     for (FormattingOption formattingOption : flags.formatting) {
       formattingOption.applyToOptions(options);
     }
-    if (flags.process_closure_primitives) {
-      options.closurePass = true;
-    }
 
+    options.closurePass = flags.process_closure_primitives;
     initOptionsFromFlags(options);
     return options;
   }
diff --git a/test/com/google/javascript/jscomp/CommandLineRunnerTest.java b/test/com/google/javascript/jscomp/CommandLineRunnerTest.java
index 0149df05..3b783ada 100644
--- a/test/com/google/javascript/jscomp/CommandLineRunnerTest.java
+++ b/test/com/google/javascript/jscomp/CommandLineRunnerTest.java
@@ -151,6 +151,13 @@ public class CommandLineRunnerTest extends TestCase {
     testSame("/** @type { not a type name } */ var x;");
   }
 
+  public void testProcessClosurePrimitives() {
+    test("var goog = {}; goog.provide('goog.dom');",
+         "var goog = {}; goog.dom = {};");
+    args.add("--process_closure_primitives=false");
+    testSame("var goog = {}; goog.provide('goog.dom');");
+  }
+
   //////////////////////////////////////////////////////////////////////////////
   // Integration tests
 
