diff --git a/src/com/google/javascript/jscomp/LiveVariablesAnalysis.java b/src/com/google/javascript/jscomp/LiveVariablesAnalysis.java
index 772d2511..e107faa9 100644
--- a/src/com/google/javascript/jscomp/LiveVariablesAnalysis.java
+++ b/src/com/google/javascript/jscomp/LiveVariablesAnalysis.java
@@ -203,8 +203,12 @@ class LiveVariablesAnalysis extends
             // for(var x in y) {...}
             lhs = lhs.getLastChild();
           }
-          addToSetIfLocal(lhs, kill);
-          addToSetIfLocal(lhs, gen);
+          if (NodeUtil.isName(lhs)) {
+            addToSetIfLocal(lhs, kill);
+            addToSetIfLocal(lhs, gen);
+          } else {
+            computeGenKill(lhs, gen, kill, conditional);
+          }
           computeGenKill(rhs, gen, kill, conditional);
         }
         return;
diff --git a/test/com/google/javascript/jscomp/LiveVariableAnalysisTest.java b/test/com/google/javascript/jscomp/LiveVariableAnalysisTest.java
index f29e5a74..1459bd49 100644
--- a/test/com/google/javascript/jscomp/LiveVariableAnalysisTest.java
+++ b/test/com/google/javascript/jscomp/LiveVariableAnalysisTest.java
@@ -206,6 +206,10 @@ public class LiveVariableAnalysisTest extends TestCase {
     assertNotLiveAfterX("var a;X:a();a(param2)", "param1");
   }
 
+  public void testExpressionInForIn() {
+    assertLiveBeforeX("var a = [0]; X:for (a[1] in foo) { }", "a");
+  }
+
   public void testArgumentsArray() {
     // Check that use of arguments forces the parameters into the
     // escaped set.
